pennylane/qnode.py

```diff
@@ -521,6 +524,16 @@ def construct(self, args, kwargs):
                     " differentiation method"
                 )
 
+        # Apply the deferred measurement principle if the device doesn't
+        # support mid-circuit measurements natively
+        # TODO: change once mid-circuit measurements are not considered as tape
+        # operations
+        if any(
+            getattr(obs, "return_type", None) == qml.operation.MidMeasure
+            for obs in self.tape.operations
+        ):
+            self._tape = qml.defer_measurements(self._tape)

```

Comment:
@antalszava would we be able to move this to `device.expand_fn`? c.f. https://github.com/PennyLaneAI/pennylane/blob/b494a2bf1b68156a9688178e1d97dc348ebf1583/pennylane/_device.py#L601

Originally we had the Hamiltonian expansion in the QNode as well, but we decided to move all expansions to the device instead

Reply:
The alternate approach is not to use deferred measurements as a fallback, and simply raise an error asking users to apply the defer transform.

Reply:
It seems that simply moving straight into `default_expand_fn` would cause the Torch and JAX differentiability to break because the underlying tensors of the tape are being unwrapped :crying_cat_face:

Reply:
oh of course - because `default_expand_fn` occurs 'inside' the interface.

@antalszava you could also try applying the transform to `circuit` inside [`device.batch_transform`](https://github.com/PennyLaneAI/pennylane/blob/b494a2bf1b68156a9688178e1d97dc348ebf1583/pennylane/_device.py#L660), before the batch transform is applied.

This should not break differentiation!

Reply:
That seems to now break differentiability with the same Torch and JAX conditional op tests raising the following error:
```python
        diff_methods = {
            idx: info["grad_method"]
            for idx, info in tape._par_info.items()  # pylint: disable=protected-access
            if idx in tape.trainable_params
        }
    
        # check and raise an error if any parameters are non-differentiable
        nondiff_params = {idx for idx, g in diff_methods.items() if g is None}
    
        if nondiff_params:
>           raise ValueError(f"Cannot differentiate with respect to parameter(s) {nondiff_params}")
E           ValueError: Cannot differentiate with respect to parameter(s) {1}

pennylane/gradients/gradient_transform.py:58: ValueError
````

Reply:
Ah 🙁 Might be important to work out what is happening here, since this breaks our recent change that only devices perform device level transforms.